File: python/anthropic_adapter.py
import time
from pathlib import Path

from anthropic import AsyncAnthropic

from engine import Adapter
from json_repair import extract_json
from models import DecomposerResponse, NodeMetadata


class AnthropicAdapter(Adapter):
    def __init__(self, api_key: str, model: str = "claude-haiku-4-5-20251001"):
        self.client = AsyncAnthropic(api_key=api_key)
        self.model = model

    async def decompose(self, topic: str, breadth: int, system_prompt: str = None) -> DecomposerResponse:
        if system_prompt:
            # Custom or persona
            if system_prompt in ("analytical", "action"):
                prompt_path = Path(__file__).parent / ".." / "protocol" / "prompts" / f"{system_prompt}.txt"
                tmpl = prompt_path.read_text(encoding="utf-8")
            else:
                tmpl = system_prompt
        else:
            # Default: analytical
            prompt_path = Path(__file__).parent / ".." / "protocol" / "prompts" / "analytical.txt"
            tmpl = prompt_path.read_text(encoding="utf-8")

        prompt = tmpl.replace("{{BREADTH}}", str(breadth)).replace("{{TOPIC}}", topic)

        start_ms = _monotonic_ms()
        response = await self.client.messages.create(
            model=self.model, max_tokens=1024, messages=[{"role": "user", "content": prompt}]
        )
        latency_ms = _monotonic_ms() - start_ms

        text = response.content[0].text
        data = extract_json(text)
        subtopics = data.get("subtopics", [])[:breadth]

        return DecomposerResponse(
            subtopics=subtopics,
            metadata=NodeMetadata(
                tokens=response.usage.input_tokens + response.usage.output_tokens,
                model=self.model,
                latency_ms=latency_ms,
            ),
        )

    def get_model_name(self) -> str:
        return self.model


def _monotonic_ms() -> int:
    return int(time.monotonic() * 1000)

File: src/anthropic_adapter.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import Anthropic from "@anthropic-ai/sdk";
import { extractJson } from "./json_repair";
import type { Adapter, DecomposerResponse } from "./types";

export class AnthropicAdapter implements Adapter {
  private client: Anthropic;
  private model: string;

  constructor(apiKey: string, model: string = "claude-haiku-4-5-20251001") {
    this.client = new Anthropic({ apiKey });
    this.model = model;
  }

  async decompose(topic: string, breadth: number, systemPrompt?: string): Promise<DecomposerResponse> {
    let tmpl: string;
    if (systemPrompt) {
      // Custom or persona-mapped prompt
      const promptFile = systemPrompt === "analytical" || systemPrompt === "action"
        ? `prompts/${systemPrompt}.txt`
        : null;
      tmpl = promptFile
        ? readFileSync(resolve(__dirname, "..", "protocol", promptFile), "utf-8")
        : systemPrompt;
    } else {
      // Default: analytical
      tmpl = readFileSync(resolve(__dirname, "..", "protocol", "prompts", "analytical.txt"), "utf-8");
    }
    const prompt = tmpl.replace("{{BREADTH}}", String(breadth)).replace("{{TOPIC}}", topic);

    const start = Date.now();
    const response = await this.client.messages.create({
      model: this.model,
      max_tokens: 1024,
      messages: [{ role: "user", content: prompt }],
    });
    const end = Date.now();

    const text = response.content[0].type === "text" ? response.content[0].text : "";

    const data = extractJson<{ subtopics: string[] }>(text);

    return {
      subtopics: data.subtopics.slice(0, breadth),
      metadata: {
        tokens: response.usage.input_tokens + response.usage.output_tokens,
        model: this.model,
        latency_ms: end - start,
      },
    };
  }

  getModelName(): string {
    return this.model;
  }
}
